python/lsst/pipe/tasks/postprocess.py

```diff
@@ -329,75 +333,359 @@ def run(self, catalog, ccdVisitId=None):
+            if config.useGlobalExternalPhotoCalib:
+                self.inputs.remove("externalPhotoCalibTractCatalog")
+            else:
+                self.inputs.remove("externalPhotoCalibGlobalCatalog")
+        else:
+            self.inputs.remove("externalPhotoCalibTractCatalog")
+            self.inputs.remove("externalPhotoCalibGlobalCatalog")
+
+
+class WriteRecalibratedSourceTableConfig(WriteSourceTableConfig,
+                                         pipelineConnections=WriteRecalibratedSourceTableConnections):
+
+    doReevaluatePhotoCalib = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc=("Add or replace local photoCalib columns from either the calexp.photoCalib or jointcal/FGCM")
+    )
+    doReevaluateSkyWcs = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc=("Add or replace local WCS columns from either the calexp.wcs or  or jointcal")
+    )
+    doApplyExternalPhotoCalib = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc=("Whether to apply external photometric calibration via an "
+             "`lsst.afw.image.PhotoCalib` object. Uses the "
+             "``externalPhotoCalibName`` field to determine which calibration "
+             "to load."),
+    )
+    doApplyExternalSkyWcs = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc=("Whether to apply external astrometric calibration via an "
+             "`lsst.afw.geom.SkyWcs` object. Uses ``externalSkyWcsName`` "
+             "field to determine which calibration to load."),
+    )
+    useGlobalExternalPhotoCalib = pexConfig.Field(
+        dtype=bool,
+        default=True,
+        doc=("When using doApplyExternalPhotoCalib, use 'global' calibrations "
+             "that are not run per-tract.  When False, use per-tract photometric "
+             "calibration files.")
+    )
+    useGlobalExternalSkyWcs = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        doc=("When using doApplyExternalSkyWcs, use 'global' calibrations "
+             "that are not run per-tract.  When False, use per-tract wcs "
+             "files.")
+    )
+
+
+class WriteRecalibratedSourceTableTask(WriteSourceTableTask):
+    """Write source table to parquet
+    """
+    _DefaultName = "writeRecalibratedSourceTable"
+    ConfigClass = WriteRecalibratedSourceTableConfig
+
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        inputs = butlerQC.get(inputRefs)
+        inputs['ccdVisitId'] = butlerQC.quantum.dataId.pack("visit_detector")
+        inputs['exposureIdInfo'] = ExposureIdInfo.fromDataId(butlerQC.quantum.dataId, "visit_detector")
+
+        if self.config.doReevaluatePhotoCalib or self.config.doReevaluateSkyWcs:
+            if self.config.doApplyExternalPhotoCalib or self.config.doApplyExternalSkyWcs:
+                inputs['exposure'] = self.attachCalibs(inputRefs, **inputs)
+
+            inputs['catalog'] = self.addCalibColumns(**inputs)
```

Comment:
Will the call work with just **inputs given that the arguments are (self, catalog, exposure, exposureIdInfo, **kwargs)? I think it usually will but do you need any protections against a case where one of catalog, exposure, or exposureIdInfo is missing from inputs? Maybe the upstream stuff is robust enough that something else would have failed already if you found yourself in such a case.

Reply:
`self.run(**inputs)` is our canonical `runQuantum`. Task methods are expected to have arguments that match the names of the input connection `Input` s.

https://pipelines.lsst.io/v/w_2022_14/modules/lsst.pipe.base/creating-a-pipelinetask.html#overriding-task-execution 

Following the pattern: data-prep methods should take  **inputs and write out new inputs. 

If `exposure` is missing for example you have an empty QG.